org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/CompilationParticipantResult.java

```diff
@@ -143,9 +173,19 @@ public void recordNewProblems(CategorizedProblem[] newProblems) {
 	this.problems = merged;
 }
 
-void reset(boolean detectedAnnotations) {
+void reset(AnnotationBinding[] annotations) {
 	// called prior to processAnnotations
-	this.hasAnnotations = detectedAnnotations;
+	this.annotations = annotations;
+	this.classfiles = null;
+	this.addedFiles = null;
+	this.deletedFiles = null;
+	this.problems = null;
+	this.dependencies = null;
+}
+
+void reset(ClassContent[] classfiles) {
```

Comment:
I haven't read it in depth yet, but I wonder if `reset(AnnotationBinding[])` should differently from `reset(ClassContent[]).

If the participant is interested in producing more artifacts based on the annotations + class content, it would need to maintain some internal state about the build context. I suspect reset(ClassContent[]) should not discard all the other state as opposed to reset(AnnotationBinding[]) which is called at the beginning of the compilation and should discard prior state.

Reply:
`reset(AnnotationBinding[])` is called *after* the compilation, and at the moment the compile method is left then (so no further processing possible), `reset(ClassContent[])` is just called after that, I just though that if when annotation processing resets the previous state, class processing should do the same, but as the results are getting out of scope afterwards it might be enough to just reset the annotations (that are only valid when calling), or one could just allow access to these as well, I would be fine with anything here :-)

One might even make the class data available already in the `processAnnotations`, just let me know.

Reply:
You're absolutely right, I misinterpreted the API - should have looked at the code first :)

> One might even make the class data available already in the processAnnotations, just let me know.

I think it does make sense. Is the idea to allow bytecode transformation from the processor?

Reply:
I need to process the classes afterwards and access the bytes, currently it is quite inconvenient as:

1. A source file can produce multiple class files, so simply take the name to find the output does not work
2. I need to know where the class files get written

Poor-Man-Soloution is to resolve the classpath in `buildFinished`, collect all output folders, read all directories and guess that if it starts with something like `SourceFileName` and ends with `.class`, then read the file contents... that seems really vastfull and error prone given that JDT already has gathered the data :-)

Reply:
True, and it wouldn't be correct either since package private classes don't need to have the same name as the source file.

So modifications to the generated bytecode are a non-goal for this?
I was thinking about the following use-case: Add a compiler participant that understands non-null annotations and adds runtime assertions to the bytecode.

Reply:
I guess the idea can be enhanced for sure, just keep in mind that currently the class-file is already written to disk, but it would be quite easy to have another method `setBytes(byte[])` that writes new content to the class-file location, if you think it is useful I can add such method as well.

Reply:
@szarnekow I now added a `setBytes` method that also removes some code duplication in JDT itself.

Reply:
@szarnekow I have now changed this so the class content is already set in the `processAnnotations` phase and not reseting the state.

Reply:
FTR: I think this is reasonable.